source/slang/slang.cpp

```diff
@@ -559,30 +561,58 @@ void CompileRequest::generateIR()
     // in isolation.
     for( auto& translationUnit : translationUnits )
     {
+        // TODO JS:
+        // This is a bit of HACK. Apparently if we call generateIRForTranslationUnit(translationUnit) twice
```

Comment:
What exactly changes between the invocations?

Anyway, I’m not really clear on why there are two *different* bottleneck steps here, rather than just upgrading the main serialization one to preserve the source location info and validate if.

Reply:
I need to look more into generateIRForTranslationUnit it may just be diagnostic output or something like that. 

It could be a singleish bottle neck if on reconstruction we could tell the deserializer what debug information to use, because there are two sources of debug information - just copying the SourceLoc (which is fast, and holds full debug 'view' info), and the truly serializable debug information which is much more complex/convoluted. 

Could use some mechanism to control what debug information is used, and in doing so perhaps have a simpler path.

That said my current command line verify isn't really supposed to be anything more than a crutch such that the path can be tested, until we can have something more complete around serialization.

Reply:
In my mind, the direct serialization of `SourceLoc` values is a stopgap, and not something we should plan to support long-term, so if we can show that the new "full" serialization logic works as intended we should be planning to switch over to it wholesale.

As I see it we are trying to "role model" how the full compilation pipeline with separate compilation will work in the future, and we just happen to be going through memory rather than the file system right now because it makes it easier to tackle serialization piece-by-piece rather than all at once. Once we have a compiler flow that includes a full/real serializaiton bottleneck I'd like to use that *exclusively* from then on, just to minimize the potential for behavior differences that would arise with having distinct paths with and without serialization.

It may actually be that at this point, with serialization of source location information in place I might argue that it is well and truly time to turn serialization of the IR on for good. That could involve making the current "bottleneck" approach on-by-default, and then later refactoring the code to rely on that on-by-default status.

Reply:
"In my mind, the direct serialization of SourceLoc values is a stopgap, and not something we should plan to support long-term, so if we can show that the new "full" serialization logic works as intended we should be planning to switch over to it wholesale."

The new debug info path results *by design* are inferior to the information of straight SourceLoc copy because new 'true' path only serializes as if there is a single 'view' moreover it is in principal an amalgamation of views (depending what parts of the view are used where by the source). 

So leaving aside a performance argument - unless we want to make 'true serialization' handle multiple views - thus I don't see SourceLoc copy as just a stop gap. That getting higher quality debug output by users could be argued more important than the (agreed, important) desire to use a single path.

That is by having the SourceLoc copy it makes it much easier argument to just switch on bottleneck by default, because nothing is lost.

Reply:
It sounds to me like we should push to make the new source location serialization path provide equivalently good output to the old path.

Maybe I didn't scan through these changes well enough (I was off on the holidays after all), but I would have assumed it wouldn't be too hard to do something like:

* Extract an array of all raw `SourceLoc` values used in a module and sort in increasing order. Reserve entry 0 for the invalid location.
* Invent "remapped" source locations for use within the serialized module that are just indices into that array
* Scan over the entries in the raw array and for each record the set of "deltas" from the previous entry: did the actual/presumed path change? the line number? the column? This can turn into a fairly compact stream of "opcodes" that define each entry in the array in terms of the predecessor.
* Store the "runs" of instructions that share a common index into the array of source locations.

Am I missing something here, because this doesn't seem like it needs to be very complicated in practice. I'm kind of worried that having the intermediate data structure between the in-memory IR and the serialized byte stream is ending up complicating the process.

Reply:
It sounds like the only missing feature then is to support multiple `SourceView`s, and whatever complications that brings with it. I'm not completely clear on how `SourceView`s work right now, or how any of this relates to explicit `#line` directives (I know you revamped that code and I don't recall the details of how the current approach works).

My preference for a POR would be that we will drop serialization/deserialization of "raw" source locations in favor of this new path once we decide it is a Good Enough, and then any of these added details would amount to feature requests for the serialization path to handle more complex cases robustly. The decision about what is Good Enough would probably be up to you, since you know the status of all the pieces better than I do.

And FWIW all of this really comes down to what `#line` directives we emit in the generated HLSL/GLSL output (and eventually any debug info we emit to SPIR-V), which in turn only affects (1) error messages when the downstream compiler fails, and (2) step-through debugging when using tools that understand the debug info.

In practice any case where the downstream compiler fails (1) is a bug on our part, and so the source location we provide in those cases is really just a quality-of-implementation thing for a case where we are failing anyway. In turn, item (2) is kind of a mess because in practice most GPU debugging tools for graphics don't do a great job with `#line` directive anyway (they completely trip up the GPU debugger built into Visual Studio, and last time I tried nSight it didn't provide the best experience...).